src/node_encoding.cc

```diff
@@ -0,0 +1,88 @@
+#include "env-inl.h"
+#include "node.h"
+#include "node_external_reference.h"
+#include "node_internals.h"
+#include "util-inl.h"
+#include "v8-fast-api-calls.h"
+#include "v8.h"
+
+namespace node {
+
+using v8::ArrayBuffer;
+using v8::BackingStore;
+using v8::CFunction;
+using v8::Context;
+using v8::FastApiTypedArray;
+using v8::FastOneByteString;
+using v8::Isolate;
+using v8::Local;
+using v8::Object;
+using v8::String;
+using v8::Uint8Array;
+using v8::Value;
+
+namespace encoding_methods {
+
+void EncodeUtf8(const v8::FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  CHECK_GE(args.Length(), 2);
+  CHECK(args[0]->IsString());
+  CHECK(args[1]->IsUint8Array());
+
+  Local<String> str = args[0].As<String>();
+  Local<Uint8Array> out = args[1].As<Uint8Array>();
+  Local<ArrayBuffer> buf = out->Buffer();
+  size_t out_length = out->ByteLength();
+  char* write_result = static_cast<char*>(buf->Data()) + out->ByteOffset();
+
+  str->WriteUtf8(isolate,
+                 write_result,
+                 out_length,
+                 nullptr,
+                 String::NO_NULL_TERMINATION | String::REPLACE_INVALID_UTF8);
+}
+
+void latin1_to_utf8(const char* str, char* c) {
+  for (; *str; ++str) {
+    if (*str & 0x80) {
+      *c++ = *str;
+    } else {
+      *c++ = (char)(0xc0 | (unsigned)*str >> 6);
+      *c++ = (char)(0x80 | (*str & 0x3f));
+    }
+  }
+  *c++ = '\0';
+}
+
+void FastEncodeUtf8(Local<Value> receiver,
+                    const FastOneByteString& source,
+                    const FastApiTypedArray<uint8_t>& output) {
+  uint8_t* storage;
+  auto is_available = output.getStorageIfAligned(&storage);
+  CHECK(is_available);
+  latin1_to_utf8(source.data, reinterpret_cast<char*>(storage));
```

Comment:
As far as I can tell, the default implementation is safe when the size of the output buffer is too small, but the fast implementation is unsafe. Such discrepancies between the slow and fast variant are extremely risky because they are non-obvious and future uses of this API might rely on safety assumptions that are violated in the fast path.

Reply:
I force pushed with the new changes, that reduces the cost of the javascript execution. I want to make sure that the fast path is safe too with no null termination. Do you have any recommendations?